Implement the ability to choose specific sounds from SoundFonts

cbjeukendrup opened the pull request:
Resolves: #17233 

Fully implements the design from the "Mixer" Figma file

Eism commented on src/framework/audio/internal/soundfontrepository.cpp:
```diff
@@ -54,18 +68,86 @@ void SoundFontRepository::loadSoundFontPaths()
+
+    loader = new_fluid_defsfloader(settings);
+    if (!loader) {
+        goto cleanup_settings;
```
please don't use goto

Eism commented on src/playback/view/internal/inputresourceitem.cpp:
```diff
@@ -188,22 +230,173 @@ QVariantMap InputResourceItem::buildVstMenuItem(const ResourceByVendorMap& resou
 
 QVariantMap InputResourceItem::buildSoundFontsMenuItem(const ResourceByVendorMap& resourcesByVendor) const
```
I think we can split this method into several

Eism commented on src/playback/view/internal/msbasicpresetscategories.h:
```diff
@@ -0,0 +1,521 @@
+};
+
+// Based on https://docs.google.com/spreadsheets/d/1SwqZb8lq5rfv5regPSA10drWjUAoi65EuMoYtG-4k5s/edit#gid=1502434630
+static const std::vector<MsBasicItem> msBasicPresetCategories {
```
maybe it's better to put it in a JSON file and read it when the program starts

shoogle replied:
Reading from a file would be good because at some point we want to allow users to create custom categories for other SoundFonts besides MS Basic.

RomanPudashkin replied:
Casper's solution also has its advantages: we won't need to read this info from the file (which is obviously faster and safer). I think we can keep this solution for now and improve it later if the need arises

cbjeukendrup replied:
I also seem to recall that Vasily has a preference for putting things statically in code rather than reading from files, for performance reasons (but I believe that was actually in the context of `instruments.xml`)

shoogle replied:
It's certainly quicker and easier to load from the source code in this case.

In the case of instruments, I never understood the arguments since we are still reading from `instruments.xml` anyway, but I would be keen to learn. Perhaps it's because some of the data is shared with the Muse Sounds guys and Vasily didn't want them to have to write a parser?

Of course we could always use the spreadsheet to generate C++ code rather than an XML document and then we would get the best of both worlds.

shoogle commented:
The expressive instruments (e.g. "Flute Expr.") don't seem to make any sound.

Also, there is a strange gap in the Strings menu which appears when I scroll. I think this is because the menu is too big to fit on my screen, so perhaps out of scope for this PR?

![image](https://github.com/musescore/MuseScore/assets/11011881/d4b2b922-5025-4faf-a52f-8d4b0663885a)

On option would be to subdivide the strings into:

```
Strings/Single
Strings/Ensemble
```
Or perhaps:

```
Strings/Violins
Strings/Violas
Strings/Celli
Strings/Basses
Strings/Other
```

MarcSabatella commented:
> The expressive instruments (e.g. "Flute Expr.") don't seem to make any sound.

Do we even still need those separate versions?  They were necessary due to the original implementation of single note dynamics for 3.1 (responding to CC vs velocity to control volume), but I'm not sure how the new system works.

RomanPudashkin commented on src/framework/audio/internal/audioconfiguration.cpp:
```diff
@@ -20,17 +20,14 @@
-#include "log.h"
 
 //TODO: remove with global clearing of Q_OS_*** defines
 #include <QtGlobal>
```
Do we still need it?

cbjeukendrup replied:
Yes, Q_OS_WASM is still used in this file

RomanPudashkin commented on src/framework/audio/internal/soundfontrepository.cpp:
```diff
@@ -21,6 +21,17 @@
+extern "C" {
+#endif
+
+#include <sfloader/fluid_sfont.h>
```
Fluid-related code should be in
src/framework/audio/internal/synthesizers/fluidsynth

We should also not mix low-level Fluid logic with high-level logic (questions for the user) in the same class

cbjeukendrup replied:
I'm thinking about moving all soundfont related stuff to a new `soundfont` module. That might make some things a bit nicer and more natural. What do you think about that idea?

RomanPudashkin replied:
I don't think we need this module, at least for now

RomanPudashkin commented on src/framework/audio/internal/synthesizers/fluidsynth/fluidresolver.cpp:
```diff
@@ -72,27 +72,62 @@ AudioResourceMetaList FluidResolver::resolveResources() const
+            chooseAutomaticMeta.vendor = FLUID_VENDOR_NAME;
+            chooseAutomaticMeta.attributes = {
+                { u"playbackSetupData", mpe::GENERIC_SETUP_DATA_STRING },
+                { u"soundFontName", String::fromStdString(name) }
```
It's better to declare global static strings and use them instead of raw strings

RomanPudashkin commented on src/framework/audio/internal/soundfontrepository.cpp:
```diff
@@ -54,18 +68,86 @@ void SoundFontRepository::loadSoundFontPaths()
+            SoundFontPreset preset;
+            preset.program = midi::Program(bank, program);
+            preset.name = name;
+            meta.presets.push_back(preset);
```
Please use emplace_back + std::move

RomanPudashkin commented on src/framework/audio/internal/soundfontrepository.cpp:
```diff
@@ -54,18 +68,86 @@ void SoundFontRepository::loadSoundFontPaths()
+            meta.presets.push_back(preset);
+        }
+
+        m_soundFonts.insert_or_assign(path, meta);
```
Please use std::move

RomanPudashkin commented on src/framework/audio/internal/soundfontrepository.cpp:
```diff
@@ -54,18 +68,86 @@ void SoundFontRepository::loadSoundFontPaths()
 }
 
-Notification SoundFontRepository::soundFontPathsChanged() const
+SoundFontsMap SoundFontRepository::soundFonts() const
```
const SoundFontsMap&

cbjeukendrup commented:
@shoogle The strange gap consisted of five "separator lines" put together 😄. The reason that separator lines appeared, was that there were some items in the list where the title was empty, which results in a separator line. I've made the code better resistant to edge cases like empty preset names, but also fixed the root cause: in the preset categories spreadsheet, there are five `Pizz. Expr` presets, but those don't exist in the SoundFont. So I've removed those from the code (but not from the spreadsheet).

cbjeukendrup commented:
And we can make further adjustments to the categorisation/order relatively painlessly; I've found a relatively convenient way to regenerate the code. See one of the commit messages, where I described how.

shoogle commented:
> See one of the commit messages, where I described how.

It's cool that you wrote a script to do it! I don't think a [commit message](https://github.com/musescore/MuseScore/pull/18976/commits/c2ebcf43faddebbee377504ab3b7647c82946453) is the best place for it though. People will never think to read the commit messages before they go ahead and adjust the order manually. Also, the format of the JSON might change in future but you won't be able to edit the commit message after the fact.

You could put the code in the repository as a `.py` file, or as a `/* multiline comment */` in the C++ file. But if you don't want to do that, simply pasting it into your PR description here on GitHub is probably as good a place as any.

RomanPudashkin commented on src/framework/audio/internal/soundfontrepository.cpp:
```diff
@@ -54,18 +58,45 @@ void SoundFontRepository::loadSoundFontPaths()
+    RetVal<SoundFontMeta> meta = FluidSoundFontParser::parseSoundFont(path);
+
+    if (!meta.ret) {
+        LOGW() << "Failed parse SoundFont presets for " << path << ": " << meta.ret.toString();
```
Looks more like LOGE

RomanPudashkin commented on src/framework/audio/internal/synthesizers/fluidsynth/fluidsoundfontparser.cpp:
```diff
@@ -0,0 +1,93 @@
+        meta.presets.emplace_back(std::move(preset));
+    }
+
+    return RetVal<SoundFontMeta>::make_ok(std::move(meta));
```
This move will not have the desired effect :)

zacjansheski commented:
Pre-installed soundfonts get new sub-menus and "Choose automatically" options in the sub-menus

Master:
<img width="522" alt="Screen Shot 2023-08-14 at 8 50 06 AM" src="https://github.com/musescore/MuseScore/assets/69917893/c7908dea-15b3-4d4d-81da-598333f31812">

PR

<img width="789" alt="Screen Shot 2023-08-14 at 8 49 00 AM" src="https://github.com/musescore/MuseScore/assets/69917893/91b1f213-c97b-4fcc-ade5-66dcae689a1c">

zacjansheski commented:
MS crashed while I scrolled through the soundfonts.  Not sure if there is anything useful here:

[Soundfont_Crash.zip](https://github.com/musescore/MuseScore/files/12334714/Soundfont_Crash.zip)

zacjansheski commented:
Is it expected that "choose automatically" works for third party soundfonts at this point?
Could/should it be only an option for MS Basic?

zacjansheski commented:
![20230814_133344.jpg](https://github.com/musescore/MuseScore/assets/69917893/a1c1b4ef-53bb-4e97-9bf0-e701bfbe5c15)

![20230814_133614.jpg](https://github.com/musescore/MuseScore/assets/69917893/6ad12d5e-b710-4056-8eb7-c5803c10af39)

Menu sometimes stuck on Ubuntu.  Trying to find a reliable way to reproduce.

zacjansheski commented:
[MenuStuck2.zip](https://github.com/musescore/MuseScore/files/12339472/MenuStuck2.zip)

Not sure I can provide a meaningful log of the menu getting frozen on Ubuntu.  Let me know if there is anything else I can do to help here.

cbjeukendrup commented:
About the "Choose automatically" option: it seemed to be explicitly specified in the design, but I would agree to suggest that we only show it for MS Basic. For other SoundFonts, it will use exactly the same mapping as for MS Basic, which almost never makes sense. And when this mapping leads to a preset that does not exist in the SoundFont in question, Fluid falls back to the first preset in the SoundFont (that's why Marc's single-preset SoundFonts are working). 
Since there is no standard for SoundFonts (and if a particular SoundFont follows a certain standard, there is no way to detect that), it is impossible to implement any auto-selection behaviour for other SoundFonts. 

There's one more design optimisation that we could make: if a SoundFont contains only one bank, we could show the sounds directly, instead of showing "Bank 0" first. @jessjwilliamson would that make sense to you?

cbjeukendrup commented:
And about those menus that get stuck, that might not be related to this PR, since it's a UI problem and this PR does not directly touch the UI components. I seem to recall that I've seen it before on my Mac, but also without a reliable way to reproduce. I'll try if I can do anything about it, but perhaps we should not let that block this PR.

MarcSabatella commented:
MS Basic uses the General MIDI standard, as do many of the most popular other soundfonts people tend to use.  So the same mapping should be expected to join very many cases.  And in MU3, you didn't need to do anything to take advantage of this - new soundfonts automatically use the same mapping if you don't go out of your way to change it.  So please don't lose the automatic option.

zacjansheski commented:
Tested on Mac OS 11.6.5, Ubuntu 20.04.6, Windows 11
Approved

cbjeukendrup commented:
Pushed the final code review fixes. @RomanPudashkin @shoogle I've agreed with @zacjansheski that this is now ready for merging and we will look into the "expr" issue and the "menu stuck" issue later.

HeinzVo commented:
Selecting instruments now works (except the exp. variants), but ornaments play unhearable very much too fast and trills don't play at all. Would be nice, if that could be fixed too when using soundfonts.

shoogle commented:
@HeinzVo, please open an issue to report any problems you are having with sounds.
